### ISSUE TITLE
detectron2.engine.defaults.DefaultTrainer.resume_or_load() didn't work

### ISSUE THREAD

**Issue by ge-li** [facebookresearch/detectron2#222]
The DefaultTrainer.resume_or_load(resume=False) method didn't work when I try to restart training only from the previous weight and set a different learning rate.

#### First run

```
from detectron2.engine import DefaultTrainer
from detectron2.config import get_cfg

cfg = get_cfg()
cfg.merge_from_file("./detectron2_repo/configs/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml")
cfg.DATASETS.TRAIN = ("balloon/train",)
cfg.DATASETS.TEST = ()   # no metrics implemented for this dataset
cfg.DATALOADER.NUM_WORKERS = 2
cfg.MODEL.WEIGHTS = "detectron2://COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x/137849600/model_final_f10217.pkl"  # initialize from model zoo
cfg.SOLVER.IMS_PER_BATCH = 2
cfg.SOLVER.BASE_LR = 0.00025
cfg.SOLVER.MAX_ITER = 1200    # 1200 iterations reach BASE_LR=0.00025 and stay
cfg.MODEL.ROI_HEADS.BATCH_SIZE_PER_IMAGE = 128   # faster, and good enough for this toy dataset
cfg.MODEL.ROI_HEADS.NUM_CLASSES = 1  # only has one class (ballon)
os.makedirs(cfg.OUTPUT_DIR, exist_ok=True)
trainer = DefaultTrainer(cfg) 
trainer.resume_or_load(resume=False)
trainer.train()
```

#### Logs:

> ...
> [11/02 22:57:59 d2.engine.train_loop]: Starting training from iteration 0
> [11/02 22:58:05 d2.utils.events]: eta: 0:05:41  iter: 19  total_loss: 2.133  loss_cls: 0.689  loss_box_reg: 0.674  loss_mask: 0.677  loss_rpn_cls: 0.020  loss_rpn_loc: 0.007  time: 0.2890  data_time: 0.0034  lr: 0.000005  max_mem: 2657M
> [11/02 22:58:11 d2.utils.events]: eta: 0:05:31  iter: 39  total_loss: 1.904  loss_cls: 0.660  loss_box_reg: 0.521  loss_mask: 0.649  loss_rpn_cls: 0.026  loss_rpn_loc: 0.008  time: 0.2818  data_time: 0.0015  lr: 0.000010  max_mem: 2657M
> ...
> [11/02 23:03:48 d2.utils.events]: eta: 0:00:06  iter: 1179  total_loss: 0.223  loss_cls: 0.039  loss_box_reg: 0.096  loss_mask: 0.058  loss_rpn_cls: 0.001  loss_rpn_loc: 0.005  time: 0.2948  data_time: 0.0017  lr: 0.000250  max_mem: 2762M
> [11/02 23:03:55 d2.utils.events]: eta: 0:00:00  iter: 1199  total_loss: 0.227  loss_cls: 0.030  loss_box_reg: 0.104  loss_mask: 0.058  loss_rpn_cls: 0.000  loss_rpn_loc: 0.005  time: 0.2951  data_time: 0.0016  lr: 0.000250  max_mem: 2762M
> [11/02 23:03:55 d2.engine.hooks]: Overall training speed: 1197 iterations in 0:05:53 (0.2953 s / it)
> [11/02 23:03:55 d2.engine.hooks]: Total training time: 0:05:54 (0:00:00 on hooks)

#### Now try to alter the learning rate from 0.00025 to 0.00005. This didn't work! The model started from iteration 1200 and ends immediately.

```
cfg.SOLVER.BASE_LR = 0.00005 # 1/5 of last base_lr
cfg.SOLVER.MAX_ITER = 200 # another 200 iterations 
cfg.MODEL.WEIGHTS = "output/model_final.pth" # use previous weights
trainer = DefaultTrainer(cfg) 
trainer.resume_or_load(resume=False) # This didn't work! The model started from iteration 1200 and ends immediately.
trainer.train()
```

#### Logs:

> ...
> [11/02 23:04:59 d2.data.build]: Using training sampler TrainingSampler
> [11/02 23:04:59 d2.engine.train_loop]: Starting training from iteration 1200
> [11/02 23:04:59 d2.engine.hooks]: Total training time: 0:00:00 (0:00:00 on hooks)

## Expected behavior

The new trainer use previous weight "output/model_final.pth" and new base learning rate = 0.00005, start training from a new round of iteration 0 to 200.

## Environment

> ------------------------  --------------------------------------------------
> sys.platform              linux
> Python                    3.7.4 (default, Aug 13 2019, 20:35:49) [GCC 7.3.0]
> Numpy                     1.17.2
> Detectron2 Compiler       GCC 7.4
> Detectron2 CUDA Compiler  10.1
> DETECTRON2_ENV_MODULE     <not set>
> PyTorch                   1.3.0
> PyTorch Debug Build       False
> torchvision               0.4.1a0+d94043a
> CUDA available            True
> GPU 0                     GeForce GTX 1080 Ti
> CUDA_HOME                 /usr/local/cuda-10.1
> NVCC                      Cuda compilation tools, release 10.1, V10.1.243
> Pillow                    6.2.0
> cv2                       4.1.1
> ------------------------  --------------------------------------------------
> PyTorch built with:
>   - GCC 7.3
>   - Intel(R) Math Kernel Library Version 2019.0.4 Product Build 20190411 for Intel(R) 64 architecture applications
>   - Intel(R) MKL-DNN v0.20.5 (Git Hash 0125f28c61c1f822fd48570b4c1066f96fcb9b2e)
>   - OpenMP 201511 (a.k.a. OpenMP 4.5)
>   - NNPACK is enabled
>   - CUDA Runtime 10.1
>   - NVCC architecture flags: -gencode;arch=compute_35,code=sm_35;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_50,code=compute_50
>   - CuDNN 7.6.3
>   - Magma 2.5.1
>   - Build settings: BLAS=MKL, BUILD_NAMEDTENSOR=OFF, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Wno-stringop-overflow, DISABLE_NUMA=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=True, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF, 
>

**Comment by ppwwyyxx** [comment]
Please include details about the problem following the [issue template](https://github.com/facebookresearch/detectron2/issues/new?template=unexpected-problems-bugs.md).

**Comment by ge-li** [comment]
> Please include details about the problem following the [issue template](https://github.com/facebookresearch/detectron2/issues/new?template=unexpected-problems-bugs.md).

Thanks for replying, I updated the issue as indicated.

**Comment by ppwwyyxx** [comment]
Because you've set `cfg.MODEL.WEIGHTS = "output/model_final.pth"` and the weights contain iteration=1200.

If you want to selectively resume a part of the checkpoint (e.g., resume the weights but don't resume the iteration), you need to modify the checkpoint file.

**Comment by ge-li** [comment]
> Because you've set `cfg.MODEL.WEIGHTS = "output/model_final.pth"` and the weights contain iteration=1200.
> 
> If you want to selectively resume a part of the checkpoint (e.g., resume the weights but don't resume the iteration), you need to modify the checkpoint file.

How exactly can I do that? Or let me rephrase this, do you have any recommended procedure to resume the weights, while keeping the iteration information, that let me reduce the learning rate for the following iterations? 

The logic would be: 
```
set MAX_ITER = N + 1000 # train another 1000 iterations
set BASE_LR = BASE_LR / 10 # reduce the learning rate
...
trainer.train() # used the reduced learning rate in the following 1000 iterations (N+1 -> N+1000) 
```

The option I'm thinking about is maybe to write my own learning rate scheduler. However, I can not anticipate the proper base_learning rate for training. Is there any ad hoc method like the one I described above?

Thanks a lot!

**Comment by ppwwyyxx** [comment]
`torch.load` gives you a dict with weights, iteration, learning rate scheduler, etc. You can modify the dict to remove what you do not want to resume and `torch.save` it.

Writing a lr scheduler seems to be the easiest solution, if the existing MultiStepLR does not work for your case.
